draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,1958 @@
+    to the correct channel.
+
+  Message
+    `Enterprise Integration Patterns`_ defines a `Message`_ as  data record that
+    the messaging system can transmit through a message channel.
+
+  Command Message
+    `Enterprise Integration Patterns`_ defines a `Command Message`_ as a
+    :term:`Message` which instructs a worker to execute a task.
+
+  Event Message
+    `Enterprise Integration Patterns`_ defines an `Event Message`_ as a
+    :term:`Message` which indicates that an event has occurred.
+
+  Document Message
+    `Enterprise Integration Patterns`_ defines a `Document Message`_ as a
+    :term:`Message` containing data from a data source.
+
+  Service Activator
+    `Enterprise Integration Patterns`_ defines a `Service Activator`_ as a
+    one-way (request only) or two-way (request-reply) adapter between the
+    :term:`Message` and the service it invokes.
+    The service can be a simple as a method call.
+    The activator handles all of the messaging details and invokes the service
+    like any other client, such that the service doesn’t even know it’s being
+    invoked through messaging.
+
+  Idempotent Receiver
+    `Enterprise Integration Patterns`_ defines an `Idempotent Receiver`_ as a
+    component that can safely receive the same message multiple times
+    but will produce the same side effects when facing duplicated messages.
+
+  Message Dispatcher
+
+    `Enterprise Integration Patterns`_ defines a `Message Dispatcher`_ as a
+    component that will consume messages from a channel and distribute them to
+    performers.
+
+  Circuit Breaker
+    Martin Fowler defines a `Circuit Breaker`_ in the following fashion:
+
+      The basic idea behind the circuit breaker is very simple.
+      You wrap a protected function call in a circuit breaker object, which monitors
+      for failures.
+      Once the failures reach a certain threshold, the circuit breaker trips,
+      and all further calls to the circuit breaker return with an error,
+      without the protected call being made at all.
+      Usually you'll also want some kind of monitor alert if the circuit breaker
+      trips.
+
+  CAP Theorem
+    TODO
+
+  Availability
+    TODO
+
+  Fault Tolerance
+    TODO
+
+  Network Resilience
+    According to Wikipedia `Network Resilience`_ is:
+
+      In computer networking: resilience is the ability to provide and maintain
+      an acceptable level of service in the face of faults and challenges to
+      normal operation.”
+      Threats and challenges for services can range from simple misconfiguration
+      over large scale natural disasters to targeted attacks.
+      As such, network resilience touches a very wide range of topics.
+      In order to increase the resilience of a given communication network,
+      the probable challenges and risks have to be identified
+      and appropriate resilience metrics have to be defined
+      for the service to be protected.
+
+  Consistency
+    TODO
```

Comment:
can you provide some source where we can go through and have some insight about this?